In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
## Set path
path = r'/Users/leo/Desktop/Career Foundry/Achievement 4/Instacart Basket Analysis 30:10:25'

In [3]:
## Import Data
ords_prods_merge = pd.read_pickle(os.path.join(path,'02 - data','prepared data','ords_prods_merge.pkl'))

# If-Statements with User-Defined Functions

In [5]:
## Create subset of first 1,000,000 rows
df = ords_prods_merge[:1000000]

In [8]:
## code for price_labels
def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

In [9]:
## Applying to data frame 
df['price_range'] = df.apply(price_label, axis=1)

/var/folders/19/p3csqgws1vqgpn31c7f5fcqh0000gn/T/ipykernel_98487/2456247371.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [12]:
## count values in price range column
df['price_range'].value_counts(dropna=False)

price_range
Mid-range product    652638
Low-range product    338018
High range             9344
Name: count, dtype: int64

In [14]:
## check most expensive product
df['prices'].max()

24.5

# If-Statements with the loc() Function

In [18]:
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [16]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [17]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [19]:
## Count values in price range loc 
df['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     652638
Low-range product     338018
High-range product      9344
Name: count, dtype: int64

# Now applying loc function to my entire data frame ords_prods_merge

In [22]:
ords_prods_merge.loc[ords_prods_merge['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [28]:
ords_prods_merge.loc[(ords_prods_merge['prices'] <= 15) & (ords_prods_merge['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [29]:
ords_prods_merge.loc[ords_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [30]:
ords_prods_merge['price_range_loc'].value_counts(dropna=False)

price_range_loc
Mid-range product     21860860
Low-range product     10126321
High-range product      417678
Name: count, dtype: int64

# If-Statements with For-Loops

In [34]:
## Find highest frequency day of the week order_dow
ords_prods_merge['order_dow'].value_counts(dropna=False)

order_dow
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

# Create new column for busiest day of the week

In [9]:
result = []

for value in ords_prods_merge["order_dow"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [10]:
## Create new column for busiest day of week
ords_prods_merge['busiest_day']=result

In [11]:
## Count values
ords_prods_merge['busiest_day'].value_counts(dropna=False)

busiest_day
Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: count, dtype: int64

In [14]:
ords_prods_merge.head(3)

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   3139998      138            28          6                 11   
1   1977647      138            30          6                 17   
2    389851      709             2          0                 21   

   days_since_prior_order  First_order  add_to_cart_order  reordered _merge  \
0                     3.0        False                  5          0   both   
1                    20.0        False                  1          1   both   
2                     6.0        False                 20          0   both   

      busiest_day  
0  Regularly busy  
1  Regularly busy  
2     Busiest day

TASK 4.7

# 2. Suppose your clients have changed their minds about the labels you created in your “busiest_day” column. Now, they want “Busiest day” to become “Busiest days” (plural). This label should correspond with the two busiest days of the week as opposed to the single busiest day. At the same time, they’d also like to know the two slowest days. Create a new column for this using a suitable method.

step 1. Change busiest_day column to busiest days. 
step 2. Make busiest_days = days 0,1 (sat,sun)
step 3. Make new column slowest_days = slowest days = days 3,4 (thu,fri)

In [19]:
## Create new column busiest_days and change values to refelct two busiest days.
result = []

for value in ords_prods_merge["order_dow"]:
  if value in [0,1]:
    result.append("Busiest days")
  else:
    result.append("Regularly busy")

In [25]:
## Create new column for busiest days
ords_prods_merge['Busiest_days']=result

In [39]:
## Create new column slowest_days and change values to refelct 2 slowest days.
result_3 = []

for value in ords_prods_merge["order_dow"]:
  if value in [3,4]:
    result_3.append("Slowest days")
  else:
      result_3.append('')

In [41]:
## Create new column for Slowest days
ords_prods_merge['Slowest_days']=result_3

In [42]:
## Check columns
ords_prods_merge.head(4)

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   3139998      138            28          6                 11   
1   1977647      138            30          6                 17   
2    389851      709             2          0                 21   
3    652770      764             1          3                 13   

   days_since_prior_order  First_order  add_to_cart_order  reordered _merge  \
0                     3.0        False                  5          0   both   
1                    20.0        False                  1          1   both   
2                     6.0        False                 20          0   both   
3                     NaN         True                 10          0   both   

     Busiest_days  Slowest_days  
0  Regularly busy                
1  Regularly busy                
2    Busiest days                
3  Regularly busy  Slowest days

# 3. Check the values of this new column for accuracy. Note any observations in markdown format.

In [44]:
ords_prods_merge['Slowest_days'].value_counts(dropna = False)

Slowest_days
                24780523
Slowest days     7624336
Name: count, dtype: int64

In [45]:
24780523 + 7624336

32404859

In [46]:
len(ords_prods_merge)

32404859

1. For my new column slowest_days we have a total of 32404859 which is the same count of rows in the data set. This tells me it is accurate.
2. We can see we have a value of 7624336 slowest days which is Thursdays and Fridays. We have a value of 24780523 'blank' which are all other days.
3. Personally I would have made 1 column which give busiest days, slowest days and regular busy lables. I have followed instructions from client and produced output above.

# 4. When too many users make Instacart orders at the same time, the app freezes. The senior technical officer at Instacart wants you to identify the busiest hours of the day. Rather than by hour, they want periods of time labeled “Most orders,” “Average orders,” and “Fewest orders.” Create a new column containing these labels called “busiest_period_of_day.”


In [48]:
## Find highest frequency hour of the order_hour_day
ords_prods_merge['order_hour_of_day'].value_counts(dropna=False)

order_hour_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

## I have identified the frequency of each hour. I have split this up into the 3 lables requested by client:
most orders: hours 9-16
average orders: 17-20
fewest orders: all other orders

why?
I grouped the hours into “Most,” “Average,” and “Fewest orders” based on the frequency of orders, with 9–16 showing the highest volumes (daytime peak), 17–20 showing moderate activity (evening), and 21-8 showing the lowest volumes (overnight and early morning).

The values for each label can be modified upon request.

In [49]:
## Making lables and assigning values for new column busiest_period_of_day
result_4 = []

for value in ords_prods_merge["order_hour_of_day"]:
    if 9 <= value <= 16:          # hours 9–16 → Most orders
        result_4.append("Most orders")
    elif 17 <= value <= 20:       # hours 17–20 → Average orders
        result_4.append("Average orders")
    else:                         # all other hours → Fewest orders
        result_4.append("Fewest orders")


In [50]:
## Making new column busiest_period_of_day with above lables and values
ords_prods_merge["busiest_period_of_day"] = result_4

In [51]:
## Checking new column has been made 
ords_prods_merge.head(4)

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   3139998      138            28          6                 11   
1   1977647      138            30          6                 17   
2    389851      709             2          0                 21   
3    652770      764             1          3                 13   

   days_since_prior_order  First_order  add_to_cart_order  reordered _merge  \
0                     3.0        False                  5          0   both   
1                    20.0        False                  1          1   both   
2                     6.0        False                 20          0   both   
3                     NaN         True                 10          0   both   

     Busiest_days  Slowest_days busiest_period_of_day  
0  Regularly busy                         Most orders  
1  Regularly busy                      Average orders  
2    Busiest days                       Fewest orders  
3  Regularly busy  Slowest days           Most orders

# 5. Print the frequency for this new column.

In [52]:
## Frequency for busiest_period_of_day column 
ords_prods_merge['busiest_period_of_day'].value_counts(dropna = False)

busiest_period_of_day
Most orders       21118071
Average orders     5958617
Fewest orders      5328171
Name: count, dtype: int64

In [53]:
21118071 + 5958617 + 5328171

32404859

In [54]:
len(ords_prods_merge)

32404859

In [ ]:
## Above is a check to ensure no values have been missed. both results total the same giving me confidence in accuracy.

# 7. Export your dataframe as a pickle file (since you added new columns) and store it correctly in your “Prepared Data” folder.

In [55]:
ords_prods_merge.to_pickle(os.path.join(path,'02 - Data','prepared data','ords_prods_merge_task_4.7.pkl'))